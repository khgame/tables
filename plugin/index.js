
module.exports = {
  rows: require('./rows'),
  mark: require('./mark'),
  desc: require('./desc'),
  expand: require('./expand'),
  plain: require('./plain')
}
