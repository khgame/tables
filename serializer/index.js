const {
  readAndTranslate
} = require('../read')
const Path = require('path')
const fs = require('fs-extra')

/**
 * serialize files with selected serializers
 * @param {string} pathIn - input file path
 * @param {string} dirOut - output dir
 * @param {Object.<string,Serializer>} serializers
 */
function serialize (pathIn, dirOut, serializers) {
  let plugins = Object.values(serializers).reduce((prev, cur) => prev.concat(cur.plugins), [])
  // console.log('plugins', plugins)
  // console.dir(Object.values(serializers))
  let ret = readAndTranslate(pathIn, {
    plugins
  })
  for (let outName in serializers) {
    let fileContent = serializers[outName].file(ret, Path.parse(pathIn).name)
    fs.outputFileSync(Path.resolve(dirOut, outName), fileContent)
  }
}

module.exports = {
  serialize,
  jsonSerializer: require('./jsonSerializer').jsonSerializer,
  tsInterfaceSerializer: require('./tsInterfaceSerializer').tsInterfaceSerializer
}
