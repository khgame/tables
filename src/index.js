export * from './plugin'
export * from './utils/read'
export * from './utils/names'
export * from './serializer'
