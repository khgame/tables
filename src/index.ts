export * from './plugin'
export * from './utils/read'
export * from './utils/names'
export * from './serializer'
export * from './core/protocol'
export * from './pipeline/layers'

export { exportTablesToTs } from './outOfBox/exportTablesToTs'
