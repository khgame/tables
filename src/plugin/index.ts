export * from './rows'
export * from './erows'
export * from './colMap'
export * from './mark'
export * from './desc'
export * from './expand'
export * from './plain'
export * from './convert'
export * from './schema'

