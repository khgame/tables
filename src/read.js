const XLSX = require('js-xlsx')

function readWorkBook (path) {
  // read the raw file
  return XLSX.readFile(path)
}

function translateWorkBook (workbook, sheetName) {
  // read the sheet with the given name or the first sheet
  const targetSheetInd = workbook.SheetNames.findIndex(n => n === (sheetName || '__data'))
  const targetSheetName = workbook.SheetNames[targetSheetInd > 0 ? targetSheetInd : 0]
  const sheet = workbook.Sheets[targetSheetName]

  // create return table
  let data = {}
  let colKeys = []

  // convert data types
  for (let key in sheet) {
    if (!sheet.hasOwnProperty(key)) continue
    let splitInd = -1
    for (let i = key.length - 1; i >= 0; i--) {
      if (key[i] >= '0' && key[i] <= '9') {
        continue
      }
      splitInd = i
      break
    }
    let col = key.slice(0, splitInd + 1)
    let row = key.slice(splitInd + 1)
    if (!row) { continue }
    if (!data[row]) data[row] = {}
    let { t, v, w } = sheet[key]
    data[row][col] = { t, v, w }
    colKeys[col] = 1
  }

  let cols = Object.keys(colKeys).sort((a, b) => a.length - b.length || a.localeCompare(b)) // 1. length 2. locale
  return {
    cols,
    data,
    getValue: (table_, row_, col_) => (!table_.data[row_]) || (!table_.data[row_][col_]) ? undefined : table_.data[row_][col_].v
  }
}

function readAndTranslate (path, options) {
  let table = translateWorkBook(readWorkBook(path), options.sheetName)
  if (options.plugins) {
    table = options.plugins.reduce((accumulate, plugin) => plugin(accumulate), table)
  }
  return table
}

module.exports = {
  readWorkBook,
  translateWorkBook,
  readAndTranslate
}
