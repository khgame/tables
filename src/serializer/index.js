const {
  readAndTranslate
} = require('../utils/read')
const Path = require('path')
const fs = require('fs-extra')

/**
 * serialize files with selected serializers
 * @param {string} pathIn - input file path
 * @param {string} dirOut - output dir
 * @param {Object.<string,Serializer>} serializers
 */
function serialize (pathIn, dirOut, serializers) {
  let plugins = Object.values(serializers).reduce((prev, cur) => prev.concat(cur.plugins), [])
  // console.log('plugins', plugins)
  // console.dir(Object.values(serializers))
  let ret = readAndTranslate(pathIn, {
    plugins
  })
  for (let outName in serializers) {
    let fileContent = serializers[outName].file(ret, Path.parse(pathIn).name)
    fs.outputFileSync(Path.resolve(dirOut, outName), fileContent)
  }
}

module.exports = {
  serialize,
  jsonSerializer: require('./jsonSerializer').jsonSerializer,
  jsSerializer: require('./jsSerializer').jsSerializer,
  tsSerializer: require('./tsSerializer').tsSerializer,
  tsInterfaceSerializer: require('./tsInterfaceSerializer').tsInterfaceSerializer
}
