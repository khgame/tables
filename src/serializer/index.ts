import { registerSerializerFormat } from './core/registry'
import type { Serializer } from '../types'
import { jsonSerializer } from './formats/json'
import { jsSerializer } from './formats/js'
import { tsSerializer } from './formats/ts'
import { tsInterfaceSerializer } from './formats/tsInterface'
import { jsonxSerializer } from './formats/jsonx'
import { goSerializer } from './formats/go'
import { csharpSerializer } from './formats/csharp'

const defaultFormats: Array<[string, { suffix: string; serializer: Serializer }]> = [
  ['json', { suffix: 'json', serializer: jsonSerializer }],
  ['js', { suffix: 'js', serializer: jsSerializer }],
  ['ts', { suffix: 'ts', serializer: tsSerializer }],
  ['ts-interface', { suffix: 'ts', serializer: tsInterfaceSerializer }],
  ['jsonx', { suffix: 'json', serializer: jsonxSerializer }],
  ['go', { suffix: 'go', serializer: goSerializer }],
  ['csharp', { suffix: 'cs', serializer: csharpSerializer }]
]

for (const [name, entry] of defaultFormats) {
  registerSerializerFormat(name, entry)
}

export * from './core'
export * from './formats/json'
export * from './formats/js'
export * from './formats/ts'
export * from './formats/tsInterface'
export * from './formats/jsonx'
export * from './formats/go'
export * from './formats/csharp'
export {
  registerSerializerFormat,
  getSerializerFormat,
  listSerializerFormats,
  removeSerializerFormat
} from './core/registry'
