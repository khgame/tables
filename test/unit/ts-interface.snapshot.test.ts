import { readAndTranslate } from '../../src/utils/read'
import { tableSchema, tableConvert } from '../../src/plugin'
import { tsInterfaceSerializer } from '../../src/serializer'
import * as Path from 'path'

describe('ts-interface serializer snapshot (example.xlsx)', () => {
  it('emits interface typedefs and repo helpers', () => {
    const excel = Path.resolve(__dirname, '../../example/example.xlsx')
    const data = readAndTranslate(excel, { plugins: [tableSchema, tableConvert] })
    const code = tsInterfaceSerializer.file(data as any, 'example', '//imports', {})
    expect(code).toContain('export interface IExample')
    expect(code).toContain('export type ExampleTID = TableContext.KHTableID;')
    expect(code).toContain('export const toExampleTID = (value: string): ExampleTID => value as ExampleTID;')
    expect(code).toContain('export class ExampleRepo')
    expect(code).toContain('static fromRaw(data: ExampleRaw): ExampleRepo')
  })
})
