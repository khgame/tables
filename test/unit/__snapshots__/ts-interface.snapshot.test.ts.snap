// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ts-interface serializer snapshot (example.xlsx) matches interface shape snapshot 1`] = `
"/** this file is auto generated */
//imports
        
export type ExampleTID = string & { readonly __ExampleTID: unique symbol };

export interface IExample {
  ctype: number;
  building: number;
  level: number;
  name: string;
  upgrage: {
    to: number;
    dependency: number[];
  };
  product: Array<{
      tid: number;
      num: number;
    }>;
  cost: Array<{key: string, val: number}>;
  arr: number[];
  pair: {key: string, val: number};
  map: {key: string, val: any}[];
  nest: Array<Array<number>|boolean[][]>;
  stars: Array<number|boolean|{key: string, val: number}>;
  nestedArray: [
    {data: number},
    {data: number}|undefined,
    Array<number|{1: string}>
  ];
  ax: number|string;
}
"
`;
